doc/bgp_loading_optimization/bgp-loading-optimization-hld.md

```diff
@@ -0,0 +1,421 @@
+
+Take `orchagent` for example, a single task of `orchagent` contains three sub-tasks `pops`, `addToSync` and `doTask`, and originally `orchagent` performs the three sub-tasks in serial. A new `pops` sub-task can only begin after the previous `doTask` is finished. The proposed design utilizes a separate thread to run `pops`, which decouples the `pops` sub-task from `addToSync` and `doTask`. As the figure shows, in the new pipeline architecture, a new `pops` sub-task begins immediately when it's ready, not having to wait for the previous `addToSync` and `doTask` to finish.
+
+<figure align=center>
+    <img src="images/pipeline-timeline.png">
+    <figcaption>Figure 7. Pipeline architecture compared with the original serial architecture<figcaption>
+</figure>  
+
+#### Ring buffer for low-cost thread coordination
+Since multiple threads are employed, we take a lock-free design by using a ring buffer as an asynchronous communication channel.
+
+#### Asynchronous sairedis API usage
+Asynchronous mode `sairedis` API is used and a list of context of response pending messages is maintained on `orchagent` to process the response when its received
+
+<figure align=center>
+    <img src="images/async-sairedis2.png" width="40%" height=20%>
+    <figcaption>Figure 8. Async sairedis workflow<figcaption>
+</figure> 
+
+#### New ResponseThread in OA
+A new `ResponseThread` is used in `orchagent` to process the response when its received so that the other threads can continue processing new routing messages
+
+### Streamlining Redis I/O
+
+The optimization for `orchagent` and `syncd` can theoretically double the BGP loading performance, which makes Redis I/O performance become a new bottleneck.
+
+#### Lower frequency of the fpmsyncd flush & APPL_DB publish
+
+Instead of flushing the `pipeline` on every data arrival and propose to use a flush timer to determine the flush frequency as illustrated below.
+
+<figure align=center>
+    <img src="images/pipeline-mode.png" height=auto>
+    <figcaption>Figure 9. Proposed new BGP loading workflow<figcaption>
+</figure>  
+
+#### Disable the temporary table mechanism in APPL_DB
+
+We propose to disable the temporary/stable table behavior and keep just a single table, so that we don't need to delete the temporary and then write into a stable one, which spares much `HDEL` and `HSET` traffic.
+
+## Low-Level Implementation
+
+### Multi-threaded orchagent with a ring buffer
+
+Orchagent now runs two threads in parallel instead of a single thread.
+
+- `table->pops(entries)` executes in the master thread to maintain the time sequence
+- `Consumer->drain()` runs in a slave thread
+- `Consumer->addToSync(entries)` is run by slave, as master is assumed to be busier
```

Comment:
>slave [](http://example.com/codeflow?start=43&length=5)

Could you replace the terminology? Do you mean main thread and a new background thread?

Reply:
Thanks for pointing this out....my bad : (